Introduce remote proc trusted application

arnopo opened the pull request:
This series introduces the remoteproc Trusted Application, which enables the authentication of signed images through the scripts/sign_rproc_fw.py script. 

The remote proc TA depends on a remoteproc PTA that must implement platform-specific requirements. To achieve this, a remoteproc PTA API is also introduced.

A ELF32 parser is also introduced. This parser, inspired from the Linux remoteproc parser,  get loadable sections and the resource table specific section.

For more details a presentation is available here: [LHR23-214 OP-TEE to extend the root of trust to the coprocessor](https://resources.linaro.org/en/resource/6c5bGvZwUAjX56fvxthxds)

Related pull request:
- https://github.com/OP-TEE/optee_os/pull/6027

Code associated to the Linux Kernel is proposed as RFC for upstream:
https://lore.kernel.org/lkml/20230523091350.292221-1-arnaud.pouliquen@foss.st.com/

etienne-lms commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -0,0 +1,1079 @@
+		 * Segment is read from a non secure memory. Crosscheck it using
+		 * the hash table to verify that the segment has not been
+		 * corrupted.
+		 */
```
I think this inline comment is misplaced or misphrased. It explains why we need a hash table, which rather belongs to `remoteproc_load_segment()` function, right below.
I think here the inline comment could state that segment info are crosschecked (load addres, load size, segment size, ...) against same info saved in hash table segment (read here from secure memory).

Suggestion (IIUC):
```
		/*
		 * Segment metadata were read from a non-secure memory.
                 * Validate them using hash table data stored in secure memory.
		 */

etienne-lms commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -0,0 +1,1079 @@
+			continue;
+
+		/*
+		 * Segment is read from a non secure memory. Crosscheck it using
```
s/non secure/non-secure/  (for consistency)

etienne-lms commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -0,0 +1,1079 @@
+	 * getting the key object from secure storage.
+	 */
+
+	/* By default ask the pta to verify the signature. */
```
/pta/remoteproc PTA/

etienne-lms commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -0,0 +1,1079 @@
+			return TEE_ERROR_BAD_PARAMETERS;
+
+		if (src < ctx->fw_img ||
+		    (src + peh->phdr.p_filesz) > (ctx->fw_img + ctx->fw_img_sz))
```
should check `src + peh->phdr.p_filesz` agianst overflow?

etienne-lms replied:
ok

etienne-lms commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -0,0 +1,1079 @@
+					  PTA_RPROC_LOAD_SEGMENT_SHA256,
+					  param_types, params, NULL);
+		if (res != TEE_SUCCESS) {
+			EMSG("Fails to load segment, res = 0x%x", res);
```
`... res = %#"PRIx32, res)`
Ditto at line 747

etienne-lms commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -0,0 +1,1079 @@
+static TEE_Result remoteproc_load_elf(struct remoteproc_context *ctx)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t num_img = 0, i = 0;
```
```c
	uint8_t num_img = 0;
	uint8_t i = 0;
```
Note that `num_img` and `i` could be `unsigned int` or `size_t`.

etienne-lms commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -0,0 +1,1079 @@
+
+/* Firmware state */
+enum remoteproc_state {
+	REMOTEPROC_OFF,
```
explicit `= 0` since it is the expected default state when context is allocated (and zero initiallized)?

etienne-lms commented on lib/libutee/include/remoteproc_pta.h:
```diff
@@ -0,0 +1,166 @@
+
+/* Firmware format */
+#define PTA_RPROC_PROPRIETARY_FMT	BIT32(0)
+#define PTA_RPROC_ELF_FMT		BIT32(1)
```
Maybe sing enumerated values above and below would better clarify which capability bits are expected from `PTA_RPROC_HW_CAPABILITIES` arguments `value[1].a` and `value[2].a`.

etienne-lms commented on lib/libutee/include/remoteproc_pta.h:
```diff
@@ -0,0 +1,166 @@
+#define PTA_RPROC_FW_SECURE_COPY	BIT32(0)
+/* The platform supports load of segment with hash protection */
+#define PTA_RPROC_FW_WITH_HASH_TABLE	BIT32(1)
+/* The platform is able to change access to secure the firmware input image */
```
a bit unclear.
`The platform is can restrict access remoteproc firmware memories`?

etienne-lms reviewed (commented):
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"ta: add remote processor firmware ELF parser" and 
"ta: add the support of the start and stop command in remoteproc TA".

arnopo reviewed (commented):
Thanks @etienne-lms for the review

etienne-lms commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -540,6 +544,64 @@ static TEE_Result remoteproc_verify_firmware(struct remoteproc_context *ctx,
+				  param_types, params, NULL);
+	if (res != TEE_SUCCESS) {
+		EMSG("Failed to translate device address %#"PRIx32, da);
+		return 0;
```
maybe add an inline comment here: `/* By convention in core, physical address 0 is invalid */`.

arnopo replied:
done

etienne-lms commented on lib/libutee/include/remoteproc_pta.h:
```diff
@@ -0,0 +1,150 @@
+ *
+ * Get Platform firmware loader service capabilities.
+ *
+ * [in]  params[0].value.a:	Unique 32bit firmware identifier
```
Actually, I wonder if `Unique 32bit remote processor identifier` is better.
The TA/PTA don't really care which system is implemented in the binary, rather where it's expected to be loaded, no?

etienne-lms reviewed (commented):
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit 
"ta: add the support of the resource table in remoteproc TA".

I've added few review comments for the 2 remaining commits.

jenswi-linaro commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -94,14 +94,23 @@ enum remoteproc_img_type {
-					  U64_ADD_TO_ALIGN(_x) : (_x); })
+#define LE16_TO_CPU(x) ({		\
+		uint8_t *__x = x;	\
+		(__x[1] << 8) | __x[0]; \
```
It's confusing with yet another way of doing this. How about exporting `core/include/io.h` to `libutils` and using `get_unaligned_le*()` instead?

jenswi-linaro replied:
> A discussion was initiate with @etienne-lms on this macro. [#6027 (comment)](https://github.com/OP-TEE/optee_os/pull/6027#discussion_r1203612792) the point was that macro should support conversion from LE to LE or BE.

Except that it does a little bit more so it's doing the exact thing (minus eventual bugs) as get_unaligned_le16().

> 
> The question from Etienne was: @jforissier, @jenswi-linaro, may I ask your advice on this? Do we need such macro? Can we assume OP-TEE OS and TAs are always little endian?

No, we can't. However, we can depend on the TA devkit to be updated to cover this. So The macros TEE_U*_{FROM,TO}_BIG_ENDIAN() should be safe in that regard.

jenswi-linaro replied:
Why do you need to make unaligned accesses?

Yes, the endian conversion functions will be updated if needed.

arnopo replied:
> Why do you need to make unaligned accesses?

Not necessary, I just need to ensure a conversion from the LE to the CPU endian which can be LE or BE.

An alternative is to create a TEE_U*_FROM_LITTLE_ENDIAN in utee_define.h

```c
/* If we we're on a big endian platform we'll have to update these */
#define TEE_U64_FROM_BIG_ENDIAN(x)  TEE_U64_BSWAP(x)
#define TEE_U32_FROM_BIG_ENDIAN(x)  TEE_U32_BSWAP(x)
#define TEE_U16_FROM_BIG_ENDIAN(x)  TEE_U16_BSWAP(x)
#define TEE_U64_TO_BIG_ENDIAN(x)    TEE_U64_BSWAP(x)
#define TEE_U32_TO_BIG_ENDIAN(x)    TEE_U32_BSWAP(x)
#define TEE_U16_TO_BIG_ENDIAN(x)    TEE_U16_BSWAP(x)
+ #define TEE_U64_FROM_LITTLE_ENDIAN(x)  x
+ #define TEE_U32_FROM_LITTLE_ENDIAN(x)  x
+ #define TEE_U16_FROM_LITTLE_ENDIAN(x)  x
+ #define TEE_U64_TO_LITTLE_ENDIAN(x)  x
+ #define TEE_U32_TO_LITTLE_ENDIAN(x)  x
+ #define TEE_U16_TO_LITTLE_ENDIAN(x)  x
```

jenswi-linaro replied:
Sounds good.

jenswi-linaro commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -0,0 +1,1102 @@
+		(__x[3] << 24) | (__x[2] << 16) | (__x[1] << 8) | __x[0]; \
+	})
+
+#define U64_ADD_TO_ALIGN(s) ({					   \
```
I'm not sure what these macros are supposed to do. How about adding a comment and implementing them using `ROUNDUP()` and friends instead?

arnopo replied:
done

jenswi-linaro commented on ta/remoteproc/sub.mk:
```diff
@@ -0,0 +1,6 @@
+# REMOTEPROC TA heap size can be customized if 4kB is not enough
+CFG_REMOTEPROC_TA_HEAP_SIZE ?= (4 * 1024)
```
No, CFG variables should not change once we start to process the `sub.mk` files.

arnopo replied:
This is inspired from the PKCS11 TA : [`ta/pkcs11/sub.mk`](https://github.com/OP-TEE/optee_os/blob/master/ta/pkcs11/sub.mk#L7)

Please, could you provide some advice on how to accomplish this in a clean manner?

jenswi-linaro replied:
Ouch, I didn't see that one before. I'd prefer to put this in `mk/config.mk`.

jforissier replied:
I also prefer not setting configuration variables in the `sub.mk` files and use them only to list the files to be compiled.

On the other hand it is convenient to keep the TA config under the TA source directory (i.e., `ta/remoteproc`). How about doing that in `ta/remoteproc/user_ta.mk`? It is included here: https://github.com/OP-TEE/optee_os/blob/3.22.0/ta/mk/build-user-ta.mk#L25. We could also introduce `ta/remoteproc/config.mk` and include it before `user_ta.mk`.

jenswi-linaro replied:
It's still somewhat late in the middle of sub.mk loading for TAs. I don't suppose we need to make a big problem out of this. So if you prefer this fine with me for now.

etienne-lms reviewed (commented):
some minor comments

arnopo commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -0,0 +1,1093 @@
+	if (!params[1].memref.buffer || !params[1].memref.size)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	DMSG("Got base addr: %p size %zx", params[1].memref.buffer,
```
done

etienne-lms commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -239,8 +231,11 @@ static TEE_Result remoteproc_get_tlv(void *tlv_chunk, size_t tlv_size,
 	while (p_tlv < p_end_tlv) {
-		tlv_type = LE32_TO_CPU(p_tlv);
-		tlv_length = LE32_TO_CPU(&p_tlv[RPROC_TLV_LENGTH_OF]);
+		p_tlv_d = (uint32_t *)(void *)p_tlv;
```
I meant here:
```c
	while (p_tlv < p_end_tlv) {
		uint32_t d32 = 0;

		memcpy(&d32, p_tlv, sizeof(d32));
		tlv_type = TEE_U32_FROM_LITTLE_ENDIAN(d32);

		memcpy(&d32, p_tlv + RPROC_TLV_LENGTH_OF, sizeof(d32));
		tlv_length = TEE_U32_FROM_LITTLE_ENDIAN(d32);

		(...)

arnopo replied:
yes much better than my implementation, thanks for the proposal.

etienne-lms commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -557,7 +559,12 @@ static paddr_t remoteproc_da_to_pa(uint32_t da, size_t size,
+	pa = (paddr_t)reg_pair_to_64(params[3].value.b, params[3].value.a);
+
+	/* On success, assert that the pa address is not 0 */
+	assert(res != TEE_SUCCESS || pa != 0);
```
You can remove the test on `res`. It's always `TEE_SUCCESS` when we're here.

arnopo replied:
done

etienne-lms commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -589,6 +582,7 @@ static TEE_Result get_tlv_images_size(struct remoteproc_context *ctx,
 {
 	TEE_Result res = TEE_ERROR_GENERIC;
 	uint8_t *tlv_value = NULL;
+	uint32_t *tlv_v = NULL;
```
```patch
-	uint32_t *tlv_v = NULL;
+	uint32_t tlv_v = 0;

etienne-lms commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -0,0 +1,1096 @@
+#define FW_TLV_PTR(img, hdr)  ((img) + sizeof(*(hdr)))
+#define FW_SIGN_PTR(img, hdr) ({					 \
+		struct remoteproc_fw_hdr *__hdr = hdr;			 \
+		FW_TLV_PTR((img), __hdr) + ROUNDUP_64(__hdr->tlv_len); \
```
sorry nitpicking. As per coding style, there needs an empty line above.
```c
#define FW_SIGN_PTR(img, hdr) ({ \
		struct remoteproc_fw_hdr *__hdr = hdr; \
		\
		FW_TLV_PTR((img), __hdr) + ROUNDUP_64(__hdr->tlv_len); \
	})
#define FW_IMG_PTR(img, hdr) ({ \
		struct remoteproc_fw_hdr *___hdr = hdr; \
		\
		FW_SIGN_PTR((img), (___hdr)) + ROUNDUP_64(___hdr->sign_len); \
	})
```

etienne-lms commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -0,0 +1,1096 @@
+	})
+#define FW_IMG_PTR(img, hdr) ({						      \
+		struct remoteproc_fw_hdr *___hdr = hdr;			      \
+		FW_SIGN_PTR((img), (___hdr)) +	ROUNDUP_64(___hdr->sign_len); \
```
typo: replace the tab with a space char.

etienne-lms commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -0,0 +1,1098 @@
+	struct remoteproc_fw_hdr *hdr = fw_orig;
+	uint32_t length = 0;
+
+	if (ADD_OVERFLOW(sizeof(*hdr), ROUNDUP_64(hdr->tlv_len), &length) ||
```
Thi should check `hdr` alignment `!IS_ALIGNED_WIT(fw_orig, struct remoteproc_fw_hdr))`.

etienne-lms commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -0,0 +1,1098 @@
+	return TEE_SUCCESS;
+
+free_sec_cpy:
+	TEE_Free(ctx->sec_cpy);
```
I see that `ctx->sec_cpy` is used only in the scope of this function `remoteproc_verify_firmware()`.
I think the field could be removed from `struct remoteproc_context` and be referenced from `remoteproc_verify_firmware()` stack, passed as argument to `remoteproc_save_fw_header_and_tlvs()`, `remoteproc_verify_header()` and `remoteproc_verify_signature()`.

arnopo replied:
The secure copy contains help hash table and TLV. I implemented in this way to be able to reuse it later out of this function.
As today I have no code that reuses it ( ST downstream) I can change this as you propose.
we can come back to a context pointer later if needed.

arnopo replied:
So here if I well understood, I keep the pointer but set it to NULL before free it.
I'm not sureto  understand your point concerning `assert(ctx->sec_cpy)` as function that use the pointer are called by `remoteproc_load_fw()`.
It seems  to me some over checking. But perhaps you prefer that add it anyway?

etienne-lms replied:
At least, in your latest proposal, `ctx->sec_cpy` is cleared when the buffer is freed. So fine with me.

arnopo commented:
I rebased on master and squashed fixup before taking into account new comments from @etienne-lms

arnopo commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -135,13 +135,15 @@ struct remoteproc_fw_hdr {
-		struct remoteproc_fw_hdr *__hdr = hdr;			 \
-		FW_TLV_PTR((img), __hdr) + ROUNDUP_64(__hdr->tlv_len); \
+#define FW_SIGN_PTR(img, hdr) ({					\
+		struct remoteproc_fw_hdr *__hdr = hdr;			\
```
done

etienne-lms reviewed (commented):
Looks good to me.

`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits:
"pta: stm32mp: add new remoteproc pta API",
"libutee: add TEE_U*_FROM/TO_LITTLE_ENDIAN()" and
"ta: add the support of the start and stop command in remoteproc TA".

`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"ta: add remote processor firmware ELF parser",
"ta: add remote processor Trusted Application" and
"ta: add the support of the resource table in remoteproc TA".

arnopo commented:
> Looks good to me.
> 
> `Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits: "pta: stm32mp: add new remoteproc pta API", "libutee: add TEE_U*_FROM/TO_LITTLE_ENDIAN()" and "ta: add the support of the start and stop command in remoteproc TA".
> 
> `Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits "ta: add remote processor firmware ELF parser", "ta: add remote processor Trusted Application" and "ta: add the support of the resource table in remoteproc TA".

Done, thanks a lot for your reviews!

jforissier reviewed (commented):
Hi @arnopo,

Please see comments below. More comments...

- For "pta: stm32mp: add new remoteproc pta API", I suggest "libutee: add remoteproc PTA API"
Why did you use a stm32mp: prefix? Isn't this API generic? It looks like other platforms could re-use this API, only the implementation will be platform-specific.. If not, the file name should probably contain the platform (remoteproc_stm32mp_pta.h).
- For "libutee: add TEE_U*_FROM/TO_LITTLE_ENDIAN()"
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`
- For the following commits I also suggest changing subjects:
  "ta: add remote processor firmware ELF parser" => "ta: remoteproc: add firmware ELF parser"
  "ta: add remote processor Trusted Application " => " ta: remoteproc: add remote processor Trusted Application"
  "ta: add the support of the start and stop command in remoteproc TA" => " ta: remoteproc: add support for start and stop commands"
  "ta: add the support of the resource table in remoteproc TA" => "ta: remoteproc: add support for resource table"
  Then: 
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

etienne-lms commented on ta/remoteproc/elf_parser.c:
```diff
@@ -0,0 +1,174 @@
+	Elf32_Ehdr *ehdr = (Elf32_Ehdr *)(void *)fw;
+
+	if (!fw || !IS_ALIGNED_WITH_TYPE(fw, uint32_t)) {
+		EMSG("Invalid fw address %p", fw);
```
I could not fine definitive specification for `PRIxVA`.
I would say `PRIxVA` is rather related to `vaddr_t` (and `PRIxPTR` for `uintptr_t` despite those 2 PRIxXXX are the same).
Recent toolchain seems to accept `%p` for any pointer (`sometype *`) . I remember old toolchains warned on non-`void *` type references.

jforissier replied:
> I could not fine definitive specification for `PRIxVA`. I would say `PRIxVA` is rather related to `vaddr_t`

Right.

> (and `PRIxPTR` for `uintptr_t` despite those 2 PRIxXXX are the same). Recent toolchain seems to accept `%p` for any pointer (`sometype *`) . I remember old toolchains warned on non-`void *` type references.

Then perhaps we should always use `"0x" PRIxPTR` to print pointers?

Also, this is an EMSG(), not a debug message, so I think we should not print an address in clear as it could help an attacker defeat ASLR. Perhaps we should consider doing like in the Linux kernel and have a CFG to hide addresses by default?

etienne-lms replied:
> Perhaps we should consider doing like in the Linux kernel and have a CFG to hide addresses by default?

Sounds a good idea. I guess this would be out of the scope of this P-R.
In the meantime, @arnopo, maybe the simplest thing here is to not print the virtual address, if you agree.

arnopo replied:
I change the print to `EMSG("Invalid firmware address");`

jenswi-linaro replied:
> Then perhaps we should always use "0x" PRIxPTR to print pointers?

I'd rather keep it apart. `PRIxPA` for `paddr_t`, `PRIxVA` for `vaddr_t`, `PRIxPTR` for `uintptr_t`, and `%p` for `void *`.

jforissier replied:
> > Then perhaps we should always use "0x" PRIxPTR to print pointers?
> 
> I'd rather keep it apart. `PRIxPA` for `paddr_t`, `PRIxVA` for `vaddr_t`, `PRIxPTR` for `uintptr_t`, and `%p` for `void *`.

Thanks for clarifying. Indeed the `I` in `PRI` stands for integer so it totally makes sense.

arnopo reviewed (commented):
Hi @jforissier 

> Hi @arnopo,
> 
> Please see comments below. More comments...
> 
> * For "pta: stm32mp: add new remoteproc pta API", I suggest "libutee: add remoteproc PTA API"
>   Why did you use a stm32mp: prefix? Isn't this API generic? It looks like other platforms could re-use this API, only the implementation will be platform-specific.. If not, the file name should probably contain the platform (remoteproc_stm32mp_pta.h).

Yes  it is generic, The implementation will be platform dependent, but not the API


> * For "libutee: add TEE_U*_FROM/TO_LITTLE_ENDIAN()"
>   `Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`
> * For the following commits I also suggest changing subjects:
>   "ta: add remote processor firmware ELF parser" => "ta: remoteproc: add firmware ELF parser"
>   "ta: add remote processor Trusted Application " => " ta: remoteproc: add remote processor Trusted Application"
>   "ta: add the support of the start and stop command in remoteproc TA" => " ta: remoteproc: add support for start and stop commands"
>   "ta: add the support of the resource table in remoteproc TA" => "ta: remoteproc: add support for resource table"
>   Then:
>   `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

Thanks for the advice, I updated the commit subjects and add your acked-by on them ( I assumed that the acked-by was for all the mentioned commits).

arnopo commented:
@jforissier 
I updated patches instead of adding fixes on top of to take your comments into account. The reason is that I changed the commit subjects and created a ta/remoteproc/src subfolder. Sorry for any inconvenient for the review process

jforissier commented on ta/remoteproc/remoteproc_core.c:
```diff
@@ -0,0 +1,1103 @@
+
+		res = TEE_InvokeTACommand(pta_session, TEE_TIMEOUT_INFINITE,
+					  PTA_RPROC_SET_MEMORY,
+					  param_types, params, NULL);
```
You're right, my mistake.

jforissier commented:
- "ta: remoteproc: add firmware ELF parser":
  `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier commented:
@arnopo do you plan on introducing the platform-specific remoteproc PTA soon? I am asking because technically this PR is "reviewed enough", but I'd rather merge it at the same time when the PTA is ready. What do you think?

arnopo commented:
> @arnopo do you plan on introducing the platform-specific remoteproc PTA soon? I am asking because technically this PR is "reviewed enough", but I'd rather merge it at the same time when the PTA is ready. What do you think?

That's seems reasonable, I will prepare a new PR with the PTA and the associated driver ( probably end of next week)
One question for this PR: Should I add your `acked-by` or `reviewed-by` on https://github.com/OP-TEE/optee_os/pull/6300/commits/787a832cc6a74ecaa832a9baf83b40a645f5d8e9? or are you waiting to review the PR with the PTA first?

jforissier commented:
> > @arnopo do you plan on introducing the platform-specific remoteproc PTA soon? I am asking because technically this PR is "reviewed enough", but I'd rather merge it at the same time when the PTA is ready. What do you think?
> 
> That's seems reasonable, I will prepare a new PR with the PTA and the associated driver ( probably end of next week) 

Good :+1: 

> One question for this PR: Should I add your `acked-by` or `reviewed-by` on [787a832](https://github.com/OP-TEE/optee_os/commit/787a832cc6a74ecaa832a9baf83b40a645f5d8e9)? or are you waiting to review the PR with the PTA first?

You may add `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>` to that commit already.

arnopo commented:
> > > @arnopo do you plan on introducing the platform-specific remoteproc PTA soon? I am asking because technically this PR is "reviewed enough", but I'd rather merge it at the same time when the PTA is ready. What do you think?
> > 
> > 
> > That's seems reasonable, I will prepare a new PR with the PTA and the associated driver ( probably end of next week)
> 
> Good 👍
> 
> > One question for this PR: Should I add your `acked-by` or `reviewed-by` on [787a832](https://github.com/OP-TEE/optee_os/commit/787a832cc6a74ecaa832a9baf83b40a645f5d8e9)? or are you waiting to review the PR with the PTA first?
> 
> You may add `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>` to that commit already.

Done, 
Thanks!

arnopo commented:
Update to suppress unexpected Change-Id added in some commit messages.

arnopo commented:
thi

> This pull request has been marked as a stale pull request because it has been open (more than) 30 days with no activity. Remove the stale label or add a comment, otherwise this pull request will automatically be closed in 5 days. Note, that you can always re-open a closed issue at any time.

This PR is waiting https://github.com/OP-TEE/optee_os/pull/6470 approval.

etienne-lms reviewed (approved):
The series looks ok to me to be merged.
P-R https://github.com/OP-TEE/optee_os/pull/6470 brings a PTA implementation based on the API introduced here.

@arnopo, maybe rebase to get an CI IBART test result.

arnopo commented:
I rebased on master done and added a fixup https://github.com/OP-TEE/optee_os/pull/6300/commits/c1d24e75998c59ce092cada37be3e60edb1bd040 for review.

arnopo commented:
> I rebased on master done and added a fixup [c1d24e7](https://github.com/OP-TEE/optee_os/commit/c1d24e75998c59ce092cada37be3e60edb1bd040) for review.

My apologies, I built with an old version of OP-TEE. The fixup is invalid since `TEE_DigestDoFinal` has been updated in  185bf58cd6 ("GP131: Update TEE_{Digest,Cipher,MAC}* functions")
So no update but only a rebase on top of the main branch.

jforissier commented:
@jenswi-linaro any more comments on this? Thanks.

arnopo commented:
Rebase on top of #6470 merge, 
I removed commit dab8b5c58d ("libutee: add new remoteproc PTA API") from this PR as it as already been merged in #6470
